hw/xbox/xid-sdl.c

```diff
@@ -464,6 +467,49 @@ static void usb_xid_class_initfn(ObjectClass *klass, void *data)
     uc->handle_attach  = usb_desc_attach;
 }
 
+
+static int haptic_init(USBXIDState *s) {
+    if (SDL_HapticOpened(s-> device_index)) { return 0; }
```

Comment:
What is this check supposed to do? Shouldn't it report an error message somehow?

Reply:
`haptic_init` is called in `usb_xbox_gamepad_realize`. In `usb_xbox_gamepad_realize` it is mentioned that poorly programmed games may make this call multiple times. The `SDL_HapticOpened` check short-circuits `haptic_init` in the case it is called repeatedly on an already initialized device.

I don't mind adding a fprintf here, or an assert if we want to see if any games actually initialize controllers multiple times.

Not sure how SDL handles multiple opened haptic devices with multiple haptic effects, but I'd rather not find out.

Reply:
Thinking about it some more, maybe this check makes more sense to live in `usb_xbox_gamepad_realize` as that method has more context. Then I can keep the check in `haptic_init` but make it a printf + assert to enforce only calling it once.

Reply:
Wouldn't an already initialized device have `sdl_haptic` non-`NULL`? This check sounds more like it checks for *another* user (like another program blocking the device, or maybe a second gamepad with the same ID).

To be honest, the documentation for this function is very vague, so I might be wrong.

Reply:
From what I gathered, all this function checks for is if the `haptic_reference_count` is `> 0`. Every time `SDL_HapticOpen[FromJoystick]` is called on a device, `haptic_reference_count` is incremented. Conversely it is decremented when `SDL_HapticClose` is called. I would _think_ this is on a per-application basis.

Reply:
>  In usb_xbox_gamepad_realize it is mentioned that poorly programmed games may make this call multiple times

This is wrong.

The comment is about something else (when to update the input, which happens later - so to prevent hammering the device we do a one-time-init to prevent it). Check the QOM documentation as to when `_realize` is called by QEMU (I've put a link in another PR comment during this review).